clients/src/main/java/org/apache/kafka/common/requests/LeaderAndIsrRequest.java

```diff
@@ -184,250 +102,96 @@ public String toString() {
-                TopicPartition topicPartition = entry.getKey();
-                partitionStateData.set(TOPIC_NAME, topicPartition.topic());
-                partitionStateData.set(PARTITION_ID, topicPartition.partition());
-                entry.getValue().setStruct(partitionStateData, version);
-                partitionStatesData.add(partitionStateData);
-            }
-            struct.set(PARTITION_STATES, partitionStatesData.toArray());
-        }
-
-        List<Struct> leadersData = new ArrayList<>(liveLeaders.size());
-        for (Node leader : liveLeaders) {
-            Struct leaderData = struct.instance(LIVE_LEADERS);
-            leaderData.set(END_POINT_ID, leader.id());
-            leaderData.set(HOST, leader.host());
-            leaderData.set(PORT, leader.port());
-            leadersData.add(leaderData);
-        }
-        struct.set(LIVE_LEADERS, leadersData.toArray());
-        return struct;
+        return data.toStruct(version());
     }
 
     @Override
     public LeaderAndIsrResponse getErrorResponse(int throttleTimeMs, Throwable e) {
+        LeaderAndIsrResponseData responseData = new LeaderAndIsrResponseData();
         Errors error = Errors.forException(e);
-
-        Map<TopicPartition, Errors> responses = new HashMap<>(partitionStates.size());
-        for (TopicPartition partition : partitionStates.keySet()) {
-            responses.put(partition, error);
+        responseData.setErrorCode(error.code());
+
+        List<LeaderAndIsrPartitionError> partitions = new ArrayList<>(partitionStates().size());
+        for (LeaderAndIsrPartitionState partition : partitionStates()) {
+            partitions.add(new LeaderAndIsrPartitionError()
+                .setTopicName(partition.topicName())
+                .setPartitionIndex(partition.partitionIndex())
+                .setErrorCode(error.code()));
         }
+        responseData.setPartitionErrors(partitions);
 
         short versionId = version();
         switch (versionId) {
             case 0:
             case 1:
             case 2:
             case 3:
-                return new LeaderAndIsrResponse(error, responses);
+                return new LeaderAndIsrResponse(responseData);
             default:
                 throw new IllegalArgumentException(String.format("Version %d is not valid. Valid versions for %s are 0 to %d",
                         versionId, this.getClass().getSimpleName(), ApiKeys.LEADER_AND_ISR.latestVersion()));
         }
     }
 
     public int controllerId() {
-        return controllerId;
+        return data.controllerId();
     }
 
     public int controllerEpoch() {
-        return controllerEpoch;
+        return data.controllerEpoch();
     }
 
-    public Map<TopicPartition, PartitionState> partitionStates() {
+    @Override
+    public long brokerEpoch() {
+        return data.brokerEpoch();
+    }
+
+    public List<LeaderAndIsrPartitionState> partitionStates() {
```

Comment:
I'd prefer not to do this kind of data duplication.  Volatiles and locking are also icky.  If we just need something we can iterate over, can we just return an `Iterator<LeaderAndIsrPartitionState>`?  It should be relatively straightforward to build a composite iterator that just iterates over each sublist in turn.